### PR TITLE
CTPPS: miniAOD

### PR DESCRIPTION
**Pull request by jan-kaspar**
This PR brings CTPPS objects to the miniAOD level.

The new data format (for both tracking and timing detectors):
`DataFormats/CTPPSReco/interface/CTPPSLocalTrackLite.h`

The ED producer
`RecoCTPPS/TotemRPLocal/plugins/CTPPSLocalTrackLiteProducer.cc`
is part of the standard reconstruction sequence
`RecoCTPPS/TotemRPLocal/python/totemRPLocalReconstruction_cff.py`
and the produced data are also included in the AOD:
`RecoCTPPS/Configuration/python/RecoCTPPS_EventContent_cff.py`

The slimming config:
`PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py`
has been updated to keep the new data format.

### CODE REVIEW

File: DataFormats/CTPPSReco/interface/CTPPSLocalTrackLite.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#ifndef DataFormats_CTPPSReco_CTPPSLocalTrackLite
+#define DataFormats_CTPPSReco_CTPPSLocalTrackLite
+
+/**
+ *\brief Local (=single RP) track with essential information only.
+ **/
+struct CTPPSLocalTrackLite
+{
+  public:
+    CTPPSLocalTrackLite(uint32_t pid=0, float px=0., float pxu=-1., float py=0., float pyu=-1., float pt=0., float ptu=-1.)
```
Comment (slava77) [inline DataFormats/CTPPSReco/interface/CTPPSLocalTrackLite.h]: please define a proper default constructor (without arguments).
Are there use cases for constructors where the first 3-4 arguments are not known while the last few can have default values?

Switch to class from struct

File: DataFormats/CTPPSReco/interface/CTPPSLocalTrackLite.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      : rpId(pid), x(px), x_unc(pxu), y(py), y_unc(pyu), time(pt), time_unc(ptu)
+    {
+    }
+
+    uint32_t getRPId() const
```
Comment (slava77) [inline DataFormats/CTPPSReco/interface/CTPPSLocalTrackLite.h]: please add doxygen-compatible comments  with documentation of the member methods (member data has some comments already, nice).

File: DataFormats/CTPPSReco/src/classes_def.xml [hunk 3]
```diff
@@ -39,4 +39,11 @@
   <class name="edm::Wrapper<edm::DetSetVector<TotemRPLocalTrack::FittedRecHit>>"/>
   <class name="std::vector<edm::DetSet<TotemRPLocalTrack::FittedRecHit> >"/>
   <class name="std::vector<TotemRPLocalTrack::FittedRecHit>"/>
+
+  <class name="CTPPSLocalTrackLite"  ClassVersion="2">
+    <version ClassVersion="2" checksum="3838813906"/>
```
Comment (slava77) [inline DataFormats/CTPPSReco/src/classes_def.xml]: version 3 is preferred now for the starting point

**Comment by jan-kaspar** [comment]
@slava77 I've implemented your suggestions.

> Are there use cases for constructors where the first 3-4 arguments are not known while the last few can have default values?

Not really - the foreseen timing detectors always have some segmentation so at least coarse spatial information is always available.

**Comment by slava77** [comment]
@cmsbuild please test

It's not quite clear to me if this Lite producer will work by running on AOD inputs
@jan-kaspar can you confirm?

**Comment by slava77** [comment]
in step3 136.731

```
   [1] Running path 'dqmoffline_step'
   [2] Calling method for module TotemRPDQMSource/'totemRPDQMSource'
Exception Message:
CTPPSDetId ctor: Invalid parameters: arm=19812843 station=5 rp=2

```

I hope the next update is first tested with the standard ```runTheMatrix.py -l limited -i all ```

**Comment by jan-kaspar** [comment]
> It's not quite clear to me if this Lite producer will work by running on AOD inputs
> @jan-kaspar can you confirm?

Yes, it can. The only input to the lite producer is `DetSetVector<TotemRPLocalTrack>` which is included in the AOD.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jan-kaspar** [comment]
@slava77 I haven't yet finished with runTheMatrix - it didn't work within my workspace based on 9_0_0_pre2 and then had to leave. But I will continue now and keep you posted.

**Comment by jan-kaspar** [comment]
With `CMSSW_9_0_X_2017-01-12-2300` I could run
```
runTheMatrix.py -l limited -i all
```
yielding
```
15 15 14 12 7 1 1 1 tests passed, 0 0 0 0 0 0 0 0 failed
```

Sorry for forgetting this test in the beginning.

**Comment by dmitrijus** [comment]
+1

**Comment by slava77** [comment]
@jan-kaspar 
is this PR compatible with 80X?
If so, you may want to make a PR for 80X, in view of possible re-miniAOD

**Comment by jan-kaspar** [comment]
> is this PR compatible with 80X?
> If so, you may want to make a PR for 80X, in view of possible re-miniAOD

Yes to both. I typically issue back-ports only after the development PR has been approved - to avoid making the same corrections in multiple branches. Let me know if you think that I should do differently this time.

**Comment by slava77** [comment]
On 1/17/17 9:04 AM, jan-kaspar wrote:
>     is this PR compatible with 80X?
>     If so, you may want to make a PR for 80X, in view of possible re-miniAOD
> 
> Yes to both. I typically issue back-ports only after the development PR
> has been approved - to avoid making the same corrections in multiple
> branches. Let me know if you think that I should do differently this time.

Thanks for clarifying.

I think we can proceed as you described, make a backport after this is
signed.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17162	5853bfc
- code changes are roughly in line with the description (there are also additional technical updates in DQM to use CTPPSDetId instead of unsigned int, apparently fixing up several spots missed in #16010 )
- jenkins tests pass and comparisons show no differences (existing products and plots are not affected)
- local test with 136.725 (2015B; the same run as what's in the shortMatrix in 137.731) had CTPPS data and shows some reasonable distributions for the new products (after the validate.C script is updated)

[red: baseline; black: with this PR, using CMSSW_9_0_X_2017-01-18-1100]
all_mini_origvssign827_runmet2016bwf136p725c_ctppslocaltracklites_ctppslocaltrackliteproducer__rereco_obj_getx
all_mini_origvssign827_runmet2016bwf136p725c_ctppslocaltracklites_ctppslocaltrackliteproducer__rereco_obj_gety

The new product branch appears in RECO/AOD and in miniAOD files with a small size, as expected
```
27        CTPPSLocalTrackLites_ctppsLocalTrackLiteProducer__reRECO
```
CPU time of the ctppsLocalTrackLiteProducer is negligible.

**Comment by slava77** [comment]
@jan-kaspar 
looks like it's time for the backport.

**Comment by jan-kaspar** [comment]
@slava77 OK, I've cherry-picked the commits from this PR to our "continuous integration" branch intended for the back-port preceeding the legacy re-reco:
  https://github.com/CTPPS/cmssw/tree/ctpps_cumulative_backport_8_0_X

**Comment by slava77** [comment]
On 1/20/17 10:45 AM, jan-kaspar wrote:
> @slava77 <https://github.com/slava77> OK, I've cherry-picked the commits
> from this PR to our "continuous integration" branch intended for the
> back-port preceeding the legacy re-reco:
> https://github.com/CTPPS/cmssw/tree/ctpps_cumulative_backport_8_0_X

OK for the rereco plans.

Actually, in case we were going to have re-miniAOD release some time
next week,
I was going to suggest you add CTPPS to miniAOD to go with it.
If you are interested, the backport will need to be minimal (no detid
migration).

**Comment by davidlange6** [comment]
Hi @jan-kaspar @slava77  -it looks like this PR is causing data workflows to crash  - eg, 136.721. Could you have a look and provide a fix. Thanks @Dr15Jones

**Comment by slava77** [comment]
On 1/23/17 11:13 AM, David Lange wrote:
> Hi @jan-kaspar <https://github.com/jan-kaspar> @slava77
> <https://github.com/slava77> -it looks like this PR is causing data
> workflows to crash - eg, 136.721. Could you have a look and provide a
> fix. Thanks @Dr15Jones <https://github.com/Dr15Jones>


the crash is related to TotemRPGeometryESModule and TotemRPUVPatternFinder

https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc530/CMSSW_9_0_X_2017-01-20-2300/pyRelValMatrixLogs/run/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/step3_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2.log

**Comment by slava77** [comment]
curiously, the crash in 136.731 happened in CMSSW_9_0_X_2017-01-20-2300 
and is no longer visible in CMSSW_9_0_X_2017-01-23-1100

**Comment by davidlange6** [comment]
but to clarify there are plenty of related ones with the same error..

**Comment by jan-kaspar** [comment]
Thanks Slava for sharing the link to the log. Reading through, I can't really understand the problem. I've run the workflow 136.731 within CMSSW_9_0_X_2017-01-20-2300 on Lxplus, but can't reproduce the problem. Would you have any hints how to debug it? Thanks!

**Comment by slava77** [comment]
On 1/24/17 2:37 AM, jan-kaspar wrote:
> Thanks Slava for sharing the link to the log. Reading through, I can't
> really understand the problem. I've run the workflow 136.731 within
> CMSSW_9_0_X_2017-01-20-2300 on Lxplus, but can't reproduce the problem.
> Would you have any hints how to debug it? Thanks!

IB tests are done multithreaded

runTheMatrix.py -l 136.731 -t 4



There is a
 static bool evRotationStoreState;
which may be introducing a problem, not sure if anything else similarly
non-threadsafe can be a problem.

**Comment by slava77** [comment]
the static bool is in https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/Geometry/VeryForwardGeometryBuilder/plugins/TotemRPGeometryESModule.cc#L180

**Comment by jan-kaspar** [comment]
Thanks Slava! Finally, with option "-j 16 -t 8" I can (sometimes) reproduce the problem. But the behaviour is quite random: sometimes the segfault occurs sometimes not, the crashing module is not always the same one...

I've tried to comment out code using the static variable in "TotemRPGeometryESModule" but it doesn't solve the problem.

Are we really sure that the problem comes from this PR? To me, it seems rather unrelated...

**Comment by davidlange6** [comment]
we aren't - I can propose to back out this PR and see if the errors go away.

**Comment by davidlange6** [comment]
to follow up here - yes, reverting this pull request fixes our problems.

**Comment by dan131riley** [comment]
AFAICT there's no reason for anything in MeasuredGeometryProducer to be static.  It looks like some function got declared static for no good reason, and that led to having to make evRotationStoreState also static, but none of those need to be static, as they can't possibly be called without a "this".

**Comment by dan131riley** [comment]
There may be a conflict between TotemRPUVPatternFinder and RPCTrigger usage of Xerces, possibly a latent concurrency bug in RPCTrigger.  I've opened issue #17288

**Comment by jan-kaspar** [comment]
> to follow up here - yes, reverting this pull request fixes our problems.

Thanks David for checking this.

> AFAICT there's no reason for anything in MeasuredGeometryProducer to be static. It looks like some function got declared static for no good reason, and that led to having to make evRotationStoreState also static, but none of those need to be static, as they can't possibly be called without a "this".

Right. The entire code of MeasuredGeometryProducer is ugly and at some point I'd like to replace it with something more standard. In a short term - if we like - it can be completely commented out. In the current workflow (unlikely to change in short future), alignment corrections are only applied at analysis level. Therefore no alignment corrections during RECO and no meanigful use of MeasuredGeometryProducer.

> There may be a conflict between TotemRPUVPatternFinder and RPCTrigger usage of Xerces, possibly a latent concurrency bug in RPCTrigger. I've opened issue #17288 

Many thanks for investigating Dan! Do I interpret correctly the discussion that the problem most likely lies in RPCTrigger?

I don't quite understand the possible interference between TotemRPUVPatternFinder and RPCTrigger since I don't think that Xerces is actually run by the CTPPS code. In steps: TotemRPUVPatternFinder does not use Xerces, it only requires geometry from EventSetup:
https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/RecoCTPPS/TotemRPLocal/plugins/TotemRPUVPatternFinder.cc#L170
The geometry builder "TotemRPGeometryESModule" would lead to Xerces use if there were alignment XML files to load. But in the current configuration there are none. So the only way how TotemRPGeometryESModule could possibly lead to use of Xerces is via requesting the geometry from the DDL files. Is Xerces used for that?

Also, I don't really understand why this PR makes this problem appear. The only thing changed in TotemRPUVPatternFinder was the interpretation of detector ids - seemingly very unrelated.

**Comment by dan131riley** [comment]
> On Jan 26, 2017, at 06:05, jan-kaspar <notifications@github.com> wrote:
> 
> I don't quite understand the possible interference between TotemRPUVPatternFinder and RPCTrigger since I don't think that Xerces is actually run by the CTPPS code. In steps: TotemRPUVPatternFinder does not use Xerces, it only requires geometry from EventSetup:
> https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/RecoCTPPS/TotemRPLocal/plugins/TotemRPUVPatternFinder.cc#L170
> The geometry builder "TotemRPGeometryESModule" would lead to Xerces use if there were alignment XML files to load. But in the current configuration there are none. So the only way how TotemRPGeometryESModule could possibly lead to use of Xerces is via requesting the geometry from the DDL files. Is Xerces used for that?

Yes, Xerces is used for the DDL files.  The simplified call stack is something like

TotemRPUVPatternFinder::produce() asks for the TotemRPGeometry which calls
TotemRPGeometryESModule::produceRealTG
TotemRPGeometryESModule::produceRealGD
TotemRPGeometryESModule::produceMeasuredGD
TotemRPGeometryESModule::produceMeasuredDDCV

That finally gets the DDCompactView via XMLIdealGeometryESSource::produce which calls DDLParser::parse which then calls into Xerces to parse it.

> Also, I don't really understand why this PR makes this problem appear. The only thing changed in TotemRPUVPatternFinder was the interpretation of detector ids - seemingly very unrelated.

It probably isn't due to any change in TotemRPUVPatternFinder, but rather comes from a change in when (or if) it gets called.  If the problem is due to interference between TotemRPUVPatternFinder and RPCTrigger, then it will only happen when those initializations overlap.  So maybe some change in that PR (like CTPPSLocalTrackLiteProducer doing a "consumes<DetSetVector<TotemRPLocalTrack>>") changes the producer schedule in a way that makes it very likely that TotemRPUVPatternFinder and RPCTrigger overlap.  Looking at how that PR changed the product dependency graph might  tell us.  Concurrency bugs are fun that way...

**Pull request merged**